crates/flipperzero/examples/threads.rs

```diff
@@ -0,0 +1,72 @@
+//! Demonstrates use of threads on the Flipper Zero.
+
+#![no_main]
+#![no_std]
+
+// Required for panic handler
+extern crate flipperzero_rt;
+
+// Required for allocator
+extern crate flipperzero_alloc;
+
+extern crate alloc;
+
+use alloc::borrow::ToOwned;
+use core::time::Duration;
+
+use flipperzero::{furi::thread, println};
+use flipperzero_rt::{entry, manifest};
+
+// Define the FAP Manifest for this application
+manifest!(name = "Threads example");
+
+// Define the entry function
+entry!(main);
+
+// Entry point
+fn main(_args: *mut u8) -> i32 {
+    println!("Main app started!");
+
+    let first = thread::spawn(|| {
+        println!("First thread started!");
+        thread::sleep(Duration::from_secs(5));
+        println!("First thread finished!");
+        0
+    });
+
+    thread::sleep(Duration::from_secs(1));
+
+    let second = thread::Builder::new()
+        .name("Flipper".to_owned())
+        .expect("name is valid")
+        .spawn(|| {
+            println!("Second thread started!");
+            thread::sleep(Duration::from_secs(2));
+            println!("Second thread finished!");
+            0
+        });
+
+    for (i, thread) in [&thread::current(), first.thread(), second.thread()]
+        .into_iter()
+        .enumerate()
+    {
+        if let Some(name) = thread.name() {
+            println!("Running thread {} ({})", i, name);
+        } else {
+            println!("Running unnamed thread {}", i);
+        }
+    }
+
+    // We can either drop the `JoinHandle` and let the thread complete in the background,
+    // or we can join the thread which blocks until its completion.
+    drop(second);
+    first.join();
+
+    // If we dropped the `JoinHandle` for `first` instead of `second` (or if we dropped
+    // both) we get a null pointer dereference, because it outlives the main thread.
+    // drop(first);
+    // second.join();
```

Comment:
This is the remaining issue that `FreeMan` doesn't currently solve: we either need some way for the main thread to implicitly join on any unfinished threads, or we need those threads to crash "safely" without producing a NPD.

Reply:
Interesting, I was also able to get a `BusFault` error when triggering this. In any case, adding a `thread::sleep(Duration::from_secs(4))` after these two lines fixes the crash as expected (because the main thread waits for a total of 7 seconds, while `first` runs for 5).

Reply:
So as far as I can tell, there is no safe way to exit a thread other than returning from main (`vTaskDelete` is not exposed and any abnormal exit calls `furi_thread_catch` which triggers a "You are doing it wrong" crash).

We could probably use `furi_thread_set_state_callback` to register a `FuriThreadStateStopped` hook for the main thread that ensures it doesn't exit until there are no other threads with the same `appid` running. Since the [loader also uses the same state callback](https://sourcegraph.com/github.com/flipperdevices/flipperzero-firmware@e42aec68c533e6120f0e776e8bdc5d7d89e99c8b/-/blob/applications/services/loader/loader.c?L300), we'll probably need to call that afterwards.

You could possibly argue that this is something that the loader should do, since it's the one about to pull the rug out from all these threads.

Reply:
> We could probably use `furi_thread_set_state_callback` to register a `FuriThreadStateStopped` hook for the main thread that ensures it doesn't exit until there are no other threads with the same `appid` running. Since the [loader also uses the same state callback](https://sourcegraph.com/github.com/flipperdevices/flipperzero-firmware@e42aec68c533e6120f0e776e8bdc5d7d89e99c8b/-/blob/applications/services/loader/loader.c?L300), we'll probably need to call that afterwards.

The state callback [can only be set while the thread is stopped](https://github.com/flipperdevices/flipperzero-firmware/blob/238005890ec6a619157ac75eb60bd02dfe6c736e/furi/core/thread.c#L229-L233), and there is no API to extract the current handler.

Reply:
Sounds good!

Might also be a good idea to print a warning in the logs, just in case people wonder why their app is hanging at exit.

Reply:
Done.